src/debug/debugService.ts

```diff
@@ -0,0 +1,251 @@
+import * as vscode from "vscode";
+import * as fs from "fs";
+import * as path from "path";
+import { ChildProcess } from "child_process";
+
+import { IDebugProvider, IDockerParser } from "./debugInterfaces";
+import * as debugUtils from "./debugUtils";
+import { DockerfileParser } from "./dockerfileParser";
+import * as providerRegistry from "./providerRegistry";
+
+import * as docker from "../docker";
+import { kubeChannel } from "../kubeChannel";
+import { Kubectl } from "../kubectl";
+import { getKubeconfig } from "../kubectlUtils";
+import { shell } from "../shell";
+import { sleep } from "../sleep";
+
+export enum DockerClient {
+    docker = "docker",
+    dockerCompose = "docker-compose"
+}
+
+export interface IDebugService {
+    launchDebug(workspaceFolder: vscode.WorkspaceFolder): Promise<void>;
+    attachDebug(workspaceFolder: vscode.WorkspaceFolder, pod?: string): Promise<void>;
+}
+
+export class DebugService implements IDebugService {
+    private debugProvider: IDebugProvider;
+    private dockerParser: IDockerParser;
+
+    constructor(private readonly kubectl: Kubectl) {
+    }
+
+    /**
+     * In launch mode, build the docker image from docker file first and then run it on kubernetes cluster,
+     * after that smartly resolve the debugging info from the docker image and create port-forward, 
+     * finally start a debugger to attach to the debugging process.
+     * 
+     * Besides, when the debug session is terminated, kill the port-forward and cleanup the created kubernetes resources (deployment/pod) automatically.
+     * 
+     * @param workspaceFolder the active workspace folder.
+     */
+    public async launchDebug(workspaceFolder: vscode.WorkspaceFolder): Promise<void> {
+        if (!workspaceFolder) {
+            return;
+        }
+        // TODO: Support docker-compose.yml
+        const dockerfilePath = path.join(workspaceFolder.uri.fsPath, "Dockerfile");
+        if (!fs.existsSync(dockerfilePath)) {
+            vscode.window.showErrorMessage(`No Dockerfile found at the workspace ${workspaceFolder.name}`);
+            return;
+        }
+        this.dockerParser = new DockerfileParser(dockerfilePath);
+        this.debugProvider = await providerRegistry.getDebugProvider(this.dockerParser.getBaseImage());
+        if (!this.debugProvider) {
+            return;

```

Comment:
No need. Because `getDebugProvider()` will fall back to a dropdown list to ask user to choose the debug provider. If after that, it is still empty, ignore it.